### PR TITLE
Add Runner Tests

### PR DESCRIPTION
**Pull request by ryanewel**
*Issue #, if available:*
https://github.com/aws-robotics/ROS2-SecTest/issues/3
*Description of changes:*
- Adds tests for runner class
- Creates a mock class for PeriodicAttackComponent 
```
Test project /home/ubuntu/sec_test_ws/build/ros_sec_test
      Start  1: test_service_utils
 1/13 Test  #1: test_service_utils ...............   Passed    0.40 sec
      Start  2: test_client_utils
 2/13 Test  #2: test_client_utils ................   Passed    0.62 sec
      Start  3: test_runner
 3/13 Test  #3: test_runner ......................   Passed    0.62 sec
      Start  4: test_attack_noop
 4/13 Test  #4: test_attack_noop .................   Passed    1.23 sec
      Start  5: test_attack_resources_cpu
 5/13 Test  #5: test_attack_resources_cpu ........   Passed    1.23 sec
      Start  6: test_attack_resources_disk
 6/13 Test  #6: test_attack_resources_disk .......   Passed    1.23 sec
      Start  7: test_attack_resources_memory
 7/13 Test  #7: test_attack_resources_memory .....   Passed    1.23 sec
      Start  8: copyright
 8/13 Test  #8: copyright ........................   Passed    0.96 sec
      Start  9: cppcheck
 9/13 Test  #9: cppcheck .........................   Passed    0.97 sec
      Start 10: cpplint
10/13 Test #10: cpplint ..........................   Passed    1.54 sec
      Start 11: lint_cmake
11/13 Test #11: lint_cmake .......................   Passed    0.91 sec
      Start 12: uncrustify
12/13 Test #12: uncrustify .......................   Passed    1.18 sec
      Start 13: xmllint
13/13 Test #13: xmllint ..........................   Passed    1.19 sec

100% tests passed, 0 tests failed out of 13

Label Time Summary:
copyright     =   0.96 sec*proc (1 test)
cppcheck      =   0.97 sec*proc (1 test)
cpplint       =   1.54 sec*proc (1 test)
gmock         =   0.62 sec*proc (1 test)
gtest         =   5.94 sec*proc (6 tests)
lint_cmake    =   0.91 sec*proc (1 test)
linter        =   6.76 sec*proc (6 tests)
uncrustify    =   1.18 sec*proc (1 test)
xmllint       =   1.19 sec*proc (1 test)

Total Test time (real) =  13.32 sec
```
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by thomas-moulard** [comment]
Is this ready for review?

**Comment by ryanewel** [comment]
I just made some small tweaks, should be good to review now.

**Comment by thomas-moulard** [comment]
Thanks! It'd be nice to clean up the PR title too.

File: CMakeLists.txt [hunk 1]
```diff
@@ -120,39 +133,51 @@ install(TARGETS
...
 install(DIRECTORY examples DESTINATION share/${PROJECT_NAME})
 
 if(BUILD_TESTING)
+  find_package(ament_cmake_gtest REQUIRED)
+  find_package(ament_cmake_gmock REQUIRED)
+  find_package(ament_lint_auto REQUIRED)
+  ament_lint_auto_find_test_dependencies()
+
+  ament_find_gmock()
+  include_directories(${GMOCK_INLCUDE_DIRS})
```
Comment (thomas-moulard) [inline CMakeLists.txt]: typo - seems like we don't need this

File: CMakeLists.txt [hunk 2]
```diff
@@ -120,39 +133,51 @@ install(TARGETS
...
+  if(DEFINED GMOCK_LIBRARIES)
```
Comment (thomas-moulard) [inline CMakeLists.txt]: Why do we need this?

File: README.md [hunk 3]
```diff
@@ -36,7 +37,7 @@ source install/local_setup.sh
 
 Run the noop demo
 ```bash
-ros2 run ros_sec_test runner __params:=`ros2 pkg prefix ros_sec_test`/share/ros_sec_test/examples/params.yaml
+ros2 run ros_sec_test ros_sec_test __params:=`ros2 pkg prefix ros_sec_test`/share/ros_sec_test/examples/params.yaml
```
Comment (thomas-moulard) [inline README.md]: This seems like an unrelated change - would you mind submitting this in another PR (the doc + renaming the binary)?

File: include/ros_sec_test/attacks/periodic_attack_component.hpp [hunk 4]
```diff
@@ -36,23 +36,25 @@ namespace attacks
 class PeriodicAttackComponent : public rclcpp_lifecycle::LifecycleNode
 {
 public:
+  PeriodicAttackComponent();
```
Comment (thomas-moulard) [inline include/ros_sec_test/attacks/periodic_attack_component.hpp]: We should try to avoid changing the class interface to enable testing. Could we remove this?

File: include/ros_sec_test/attacks/periodic_attack_component.hpp [hunk 5]
```diff
@@ -36,23 +36,25 @@ namespace attacks
...
   explicit PeriodicAttackComponent(std::string node_name);
   PeriodicAttackComponent(const PeriodicAttackComponent &) = delete;
   PeriodicAttackComponent & operator=(const PeriodicAttackComponent &) = delete;
+  virtual ~PeriodicAttackComponent() = default;
 
-  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+  virtual rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
```
Comment (thomas-moulard) [inline include/ros_sec_test/attacks/periodic_attack_component.hpp]: Ditto, let's try not to modify the class for testing purposes.

https://github.com/google/googletest/blob/master/googlemock/docs/CookBook.md#mocking-nonvirtual-methods

File: package.xml [hunk 6]
```diff
@@ -24,6 +24,7 @@
   <exec_depend>ros2run</exec_depend>
 
   <test_depend>ament_cmake_gtest</test_depend>
+  <test_depend>ament_cmake_gmock</test_depend>
```
Comment (thomas-moulard) [inline package.xml]: alphasort

File: src/attacks/periodic_attack_component.cpp [hunk 7]
```diff
@@ -39,6 +39,9 @@ namespace ros_sec_test
 namespace attacks
 {
 
+PeriodicAttackComponent::PeriodicAttackComponent()
+: PeriodicAttackComponent("attack_node") {}
```
Comment (thomas-moulard) [inline src/attacks/periodic_attack_component.cpp]: This API is dangerous as we know we'll have more than one node in the process and having multiple nodes with the same name is undesirable. Let's remove this.

File: src/runner/runner.cpp [hunk 8]
```diff
@@ -22,6 +22,7 @@
 #include "lifecycle_msgs/msg/state.hpp"
 #include "rclcpp/logger.hpp"
 #include "rclcpp_lifecycle/state.hpp"
+#include "rclcpp_lifecycle/lifecycle_node.hpp"
```
Comment (thomas-moulard) [inline src/runner/runner.cpp]: alphasort

File: src/runner/runner.hpp [hunk 9]
```diff
@@ -46,12 +46,18 @@ namespace runner
 class Runner
 {
 public:
-  /// Instantiate all nodes without activating them.
+  /// Instantiate all nodes from parameter without activating them.
```
Comment (thomas-moulard) [inline src/runner/runner.hpp]: Let's remove the default constructor. We only need one.

File: src/utilities/lifecycle_service_client.cpp [hunk 10]
```diff
@@ -86,7 +86,7 @@ bool LifecycleServiceClient::configure()
 bool LifecycleServiceClient::shutdown()
 {
   return change_state(rclcpp_lifecycle::Transition(lifecycle_msgs::msg::Transition::
-           TRANSITION_DEACTIVATE));
+           TRANSITION_ACTIVE_SHUTDOWN));
```
Comment (thomas-moulard) [inline src/utilities/lifecycle_service_client.cpp]: This seems unrelated? Could this be a different PR?

File: test/include/test_utilities/periodic_attack_component_mock.hpp [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+#include "rclcpp_lifecycle/state.hpp"
+
+
+using PeriodicAttackComponent = ros_sec_test::attacks::PeriodicAttackComponent;
+using CallbackReturn = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn;
+
+class MockPeriodicAttackComponent : public PeriodicAttackComponent
+{
+public:
+  explicit MockPeriodicAttackComponent(std::string node_name)
```
Comment (thomas-moulard) [inline test/include/test_utilities/periodic_attack_component_mock.hpp]: This is the default behavior so we don't need a constructor in this class I believe.

File: test/ros_sec_test/runner/test_runner.cpp [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+using ros_sec_test::runner::Runner;
+using CallbackReturn = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn;
+using ::testing::Return;
+using ::testing::_;
+
+TEST_F(ROSTestingFixture, test_multiple_attack_lifecycle) {
+  std::vector<LifecycleNodeShPtr> nodes;
+  auto node1 = std::make_shared<MockPeriodicAttackComponent>("first_attack_node");
+  auto node2 = std::make_shared<MockPeriodicAttackComponent>("second_attack_node");
+  nodes.push_back(node1);
```
Comment (thomas-moulard) [inline test/ros_sec_test/runner/test_runner.cpp]: Nit: you could also write

```
auto node1 = std::make_shared<MockPeriodicAttackComponent>("first_attack_node");
auto node2 = std::make_shared<MockPeriodicAttackComponent>("second_attack_node");
auto nodes = {node1, node2};
Runner runner(nodes);
```

File: src/utilities/lifecycle_service_client.cpp (hunk 10 above)
Comment (ryanewel) [reply src/utilities/lifecycle_service_client.cpp]: Yep, just noticed this when writing the tests. Separate PR is here: https://github.com/aws-robotics/ROS2-SecTest/pull/15

File: test/include/test_utilities/periodic_attack_component_mock.hpp (hunk 11 above)
Comment (thomas-moulard) [reply test/include/test_utilities/periodic_attack_component_mock.hpp]: Nevermind, we can keep this.

File: src/utilities/lifecycle_service_client.cpp (hunk 10 above)
Comment (thomas-moulard) [reply src/utilities/lifecycle_service_client.cpp]: #15 is merged

File: README.md (hunk 3 above)
Comment (thomas-moulard) [reply README.md]: ok!

File: test/ros_sec_test/runner/test_runner.cpp (hunk 12 above)
Comment (thomas-moulard) [reply test/ros_sec_test/runner/test_runner.cpp]: Ok, I thought it would convert implicitly but seems like it's not working so we can keep it as it is.

Comment (thomas-moulard) [reply test/ros_sec_test/runner/test_runner.cpp]: Note: another solution would be to update the constructor to take a `std::initializer_list<rclcpp_lifecycle::LifecycleNode::SharedPtr>`.

File: src/runner/runner.hpp (hunk 9 above)
Comment (thomas-moulard) [reply src/runner/runner.hpp]: That makes sense, let's implement `Runner()` using `Runner(std::vector<...>)` to avoid duplicating the code.

File: test/ros_sec_test/runner/test_runner.cpp (hunk 12 above)
Comment (thomas-moulard) [reply test/ros_sec_test/runner/test_runner.cpp]: re-opening in case you want to try the initializer list route :)
